File: 2024/javascript/04-decoratingTheChristmasTree.js
// 04. Decorating The Christmas Tree.

/**
 * @param {number} height - Height of the tree
 * @param {string} ornament - Symbol to draw
 * @returns {string} Drawn tree
 */
function createXmasTree(height, ornament) {
  const xmasTree = [];
  const width = height * 2 - 1;

  for (let i = 1; i <= width + 1; i += 2) {
    const spaces = "_".repeat((width - i) / 2);
    const level = ornament.repeat(i);
    xmasTree.push(spaces + level + spaces);
  }

  const trunkSpaces = "_".repeat(width / 2);
  const trunk = `${trunkSpaces}#${trunkSpaces}`;
  xmasTree.push(trunk, trunk);

  return xmasTree.join("\n");
}

const tree = createXmasTree(5, "*");
console.log(tree);
/*
____*____
___***___
__*****__
_*******_
*********
____#____
____#____
*/

const tree2 = createXmasTree(3, "+");
console.log(tree2);
/*
__+__
_+++_
+++++
__#__
__#__
*/

const tree3 = createXmasTree(6, "@");
console.log(tree3);
/*
_____@_____
____@@@____
___@@@@@___
__@@@@@@@__
_@@@@@@@@@_
@@@@@@@@@@@
_____#_____
_____#_____
*/

File: 2024/python/04_decorating_the_christmas_tree.py
"""Challenge #4: 🎄Decorating The Christmas Tree."""


def create_xmas_tree(height: int, ornament: str) -> str:
    """
    Create christmas tree.

    Args:
      height (int): Height of the tree
      ornament (str): Symbol to draw

    Returns:
        str: Drawn tree
    """
    width = height * 2 - 1
    xmas_tree = [(ornament * i).center(width, "_") for i in range(1, width + 1, 2)]

    # Add the trunk
    trunk_part = "#".center(width, "_")
    xmas_tree.extend([trunk_part] * 2)

    return "\n".join(xmas_tree)


print(create_xmas_tree(5, "*"))
# ____*____
# ___***___
# __*****__
# _*******_
# *********
# ____#____
# ____#____

print(create_xmas_tree(3, "+"))
# __+__
# _+++_
# +++++
# __#__
# __#__

print(create_xmas_tree(6, "@"))
# _____@_____
# ____@@@____
# ___@@@@@___
# __@@@@@@@__
# _@@@@@@@@@_
# @@@@@@@@@@@
# _____#_____
# _____#_____
